Привет, меня зовут Артем Хуршудов. Сегодня я проверю твой проект.
<br> Дальнейшее общение будет происходить на "ты" если это не вызывает никаких проблем.
<br> Желательно реагировать на каждый мой комментарий ('исправил', 'не понятно как исправить ошибку', ...)
<br> Пожалуйста, не удаляй комментарии ревьюера, так как они повышают качество повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:

<div class="alert alert-block alert-success">
<b>Успех:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>

-------------------

Будет очень хорошо, если ты будешь помечать свои действия следующим образом:
<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> ...
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Были внесены следующие изменения ...
</div>

<font color='green'><b>Полезные (и просто интересные) материалы:</b></font> \
Для работы с текстами используют и другие подходы. Например, сейчас активно используются RNN (LSTM) и трансформеры (BERT и другие с улицы Сезам, например, ELMO). НО! Они не являются панацеей, не всегда они нужны, так как и TF-IDF или Word2Vec + модели из классического ML тоже могут справляться. \
BERT тяжелый, существует много его вариаций для разных задач, есть готовые модели, есть надстройки над библиотекой transformers. Если, обучать BERT на GPU (можно в Google Colab или Kaggle), то должно быть побыстрее.\
https://huggingface.co/transformers/model_doc/bert.html \
https://t.me/renat_alimbekov \
https://colah.github.io/posts/2015-08-Understanding-LSTMs/ - Про LSTM \
https://web.stanford.edu/~jurafsky/slp3/10.pdf - про энкодер-декодер модели, этеншены\
https://pytorch.org/tutorials/beginner/transformer_tutorial.html - официальный гайд
по трансформеру от создателей pytorch\
https://transformer.huggingface.co/ - поболтать с трансформером \
Библиотеки: allennlp, fairseq, transformers, tensorflow-text — множествореализованных
методов для трансформеров методов NLP \
Word2Vec https://radimrehurek.com/gensim/models/word2vec.html 

<font color='green'>Пример BERT с GPU:
```python
%%time
from tqdm import notebook
batch_size = 2 # для примера возьмем такой батч, где будет всего две строки датасета
embeddings = [] 
for i in notebook.tqdm(range(input_ids.shape[0] // batch_size)):
        batch = torch.LongTensor(input_ids[batch_size*i:batch_size*(i+1)]).cuda() # закидываем тензор на GPU
        attention_mask_batch = torch.LongTensor(attention_mask[batch_size*i:batch_size*(i+1)]).cuda()
        
        with torch.no_grad():
            model.cuda()
            batch_embeddings = model(batch, attention_mask=attention_mask_batch)
        
        embeddings.append(batch_embeddings[0][:,0,:].cpu().numpy()) # перевод обратно на проц, чтобы в нумпай кинуть
        del batch
        del attention_mask_batch
        del batch_embeddings
        
features = np.concatenate(embeddings) 
```
Можно сделать предварительную проверку на наличие GPU.\
Например, так: ```device = torch.device("cuda:0") if torch.cuda.is_available() else torch.device("cpu")```\
Тогда вместо .cuda() нужно писать .to(device)

Если понравилась работа с текстами, то можешь посмотреть очень интересный (но очень-очень сложный) курс лекций: https://github.com/yandexdataschool/nlp_course .
</font>

### <font color='orange'>Общее впечатление</font>
* Большое спасибо за проделанную работу. Видно, что приложено много усилий.
* Радует, что ноутбук хорошо структурирован. Приятно проверять такие работы.
* Отлично, что стоп-слова были исключены при векторизации!
* Работа получилась отличной, тебе удалось добиться достаточно хорошего качества. Поздравляю!
* Проект может быть зачтен, но я его отправлю назад, чтобы у тебя была возможность задать вопросы и внести правки, при желании. Однако, ты можешь просто вернуть проект в таком же виде и я его зачту.

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> Большое спасибо за ревью! Изменений не вносил, ссылки отдельно почитаю ;)
</div>

### <font color='orange'>Общее впечатление (ревью 2)</font>
* Проект зачтен!
* Удачи в дальнейшем обучении и следующих работах!

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Общее-впечатление" data-toc-modified-id="Общее-впечатление-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span><font color="orange">Общее впечатление</font></a></span></li><li><span><a href="#Общее-впечатление-(ревью-2)" data-toc-modified-id="Общее-впечатление-(ревью-2)-0.2"><span class="toc-item-num">0.2&nbsp;&nbsp;</span><font color="orange">Общее впечатление (ревью 2)</font></a></span></li></ul></li><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span><ul class="toc-item"><li><span><a href="#Первичный-анализ" data-toc-modified-id="Первичный-анализ-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Первичный анализ</a></span></li><li><span><a href="#Приведение-данных" data-toc-modified-id="Приведение-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Приведение данных</a></span></li></ul></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span><ul class="toc-item"><li><span><a href="#Подбор-модели" data-toc-modified-id="Подбор-модели-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Подбор модели</a></span></li><li><span><a href="#Тест-модели" data-toc-modified-id="Тест-модели-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Тест модели</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Проект для «Викишоп»

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

Для выполнения проекта применять *BERT* необязательно, но вы можете попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

## Подготовка

### Первичный анализ

In [214]:
import warnings
warnings.filterwarnings('ignore')

import pandas as pd
import numpy as np

import nltk
from nltk.stem import WordNetLemmatizer
import re
from nltk.corpus import stopwords as nltk_stopwords
from sklearn.feature_extraction.text import TfidfVectorizer

from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.metrics import f1_score

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

<div class="alert alert-block alert-success">
<b>Успех:</b> Отлично, что все импорты собраны в первой ячейке ноутбука! Если у того, кто будет запускать твой ноутбук будут отсутствовать некоторые библиотеки, то он это увидит сразу, а не в процессе!
</div>

In [215]:
df = pd.read_csv('/datasets/toxic_comments.csv')
df.info()
display(df.head())
print('Дубликатов', df.duplicated().sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159571 entries, 0 to 159570
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   text    159571 non-null  object
 1   toxic   159571 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 2.4+ MB


text  toxic
0  Explanation\nWhy the edits made under my usern...      0
1  D'aww! He matches this background colour I'm s...      0
2  Hey man, I'm really not trying to edit war. It...      0
3  "\nMore\nI can't make any real suggestions on ...      0
4  You, sir, are my hero. Any chance you remember...      0

Дубликатов 0


Датасет содержит 159,571 наблюдение. Пропусков и дубликатов нет.

<div class="alert alert-block alert-success">
<b>Успех:</b> Данные загружены корректно.
</div>

### Приведение данных

In [216]:
#text lemmatization
m = WordNetLemmatizer()
def lemmatize(text):
    lem = "".join(m.lemmatize(text)).lower()
    return lem

def clear_text(text):
    t = re.sub(r'[^a-zA-Z ]', ' ', text)
    t = t.split()
    return " ".join(t)

df['lem_text'] = df['text'].apply(clear_text).apply(lemmatize)

<div class="alert alert-block alert-success">
<b>Успех:</b> Очистка была сделана верно.
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Этот лемматизатор нужно применять к словам, а не ко всему тексту сразу.
</div>

In [217]:
#samples split
df_train, df_test = train_test_split(df, test_size=0.25, random_state=12345)

target_train = df_train['toxic']
target_test = df_test['toxic']

corpus_train = df_train['lem_text'].values.astype('U')
corpus_test = df_test['lem_text'].values.astype('U')

## Обучение

### Подбор модели

In [218]:
#vectorization
count_tf_idf = TfidfVectorizer(stop_words=stopwords)
tf_idf = count_tf_idf.fit(corpus_train)
features_train = tf_idf.transform(corpus_train)
features_test = tf_idf.transform(corpus_test)

<div class="alert alert-block alert-success">
<b>Успех:</b> Разбиение было сделано правильно. Отлично, что векторизатор был обучен только на тренировочной части данных.
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Приводить тексты к юникоду не имеет смысла, так как они все на английском.
</div>

In [219]:
%%time
#LogisticRegression
model = LogisticRegression(class_weight='balanced')
cvs = cross_val_score(model, features_train, target_train, cv=5, scoring='f1').mean()
print(f'cross val score: {cvs:.2f}')

cross val score: 0.75
CPU times: user 1min 7s, sys: 13.7 s, total: 1min 21s
Wall time: 10.6 s


<div class="alert alert-block alert-info">
<b>Совет: </b> Напомню, что внутри кросс-валидации происходит разбиение выборки на треин и валидацию. Однако, в таком случае векторизатор обучен на всей выборке, а это не совсем корректно. Для избежания такого эффекта можно использовать <a href="https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html">пайплайн</a>. <a href="https://medium.com/analytics-vidhya/ml-pipelines-using-scikit-learn-and-gridsearchcv-fe605a7f9e05">Тут</a> есть пример.
</div>

In [220]:
%%time
#DecisionTreeClassifier
cvs_max = 0
depth_best = 1
for i in range(1, 5):
    model = DecisionTreeClassifier(max_depth=i, random_state=12345, class_weight='balanced')
    cvs = cross_val_score(model, features_train, target_train, cv=5, scoring='f1').mean()
    if cvs > cvs_max:
        cvs_max = cvs
        depth_best = i
    
print(f'best depth - {depth_best}, cross val score: {cvs_max:.2f}')

best depth - 4, cross val score: 0.42
CPU times: user 2min 3s, sys: 415 ms, total: 2min 4s
Wall time: 2min 2s


In [221]:
%%time
#RandomForestClassifier
cvs_max = 0
depth_best = 1
est_best = 1
for i in range(1, 5):
    for j in range(1, 5):
        model = RandomForestClassifier(max_depth=i, n_estimators=j, random_state=12345, class_weight='balanced')
        cvs = cross_val_score(model, features_train, target_train, cv=5, scoring='f1').mean()
        if cvs > cvs_max:
            cvs_max = cvs
            depth_best = i
            est_best = j
    
print(f'best depth - {depth_best}, best estimators - {est_best}, cross val score: {cvs_max:.2f}')

best depth - 4, best estimators - 4, cross val score: 0.23
CPU times: user 12.2 s, sys: 23.7 ms, total: 12.2 s
Wall time: 12.2 s


<div class="alert alert-block alert-success">
<b>Успех:</b> Молодец, что попробовал разные модели в этом шаге!
</div>

### Тест модели

Наилучший результат показала модель логистической регрессии. Оценим ее на тестовой выборке.

In [222]:
model = LogisticRegression(class_weight='balanced')
model.fit(features_train,target_train)
predicted = model.predict(features_test)
f1 = f1_score(target_test, predicted)
print(f'f1-score: {f1:.2f}')

f1-score: 0.76


<div class="alert alert-block alert-success">
<b>Успех:</b> Тестирование было сделано верно.
</div>

## Выводы

Модель логистической регрессии для оценки эмоционального окраса выглядит наиболее оптимальной, так как предоставляет наивысший результат cross_val_score, сопровождаемый гораздо более высокой скоростью работы. 

## Чек-лист проверки

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Данные загружены и подготовлены
- [x]  Модели обучены
- [x]  Значение метрики *F1* не меньше 0.75
- [x]  Выводы написаны